### PR TITLE
NIAD-765 Added TTL mechanism

### PR DESCRIPTION
**Pull request by mmackiewicz1**

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/gp2gp/common/mongo/ttl/MongoTtlCreator.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.springframework.data.domain.Sort;
+import org.springframework.data.mongodb.core.index.Index;
+import org.springframework.data.mongodb.core.index.IndexInfo;
+import org.springframework.data.mongodb.core.index.IndexOperations;
+
+@Slf4j
+public class MongoTtlCreator extends TtlCreator {
+    public static final String TTL_INDEX_NAME = "TTL";
+    private static final String FIELD_KEY = "translationTimestamp";
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/common/mongo/ttl/MongoTtlCreator.java:16]: Our EhrExtractStatus has no translationTimestamp.

This index should be on the field "updatedAt"

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/TimeToLiveTest.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Test
+    public void When_ApplicationStarts_Expect_TtlIndexExistsForEhrExtractStatusWithValueFromConfiguration() {
+        IndexOperations indexOperations = mongoTemplate.indexOps(EhrExtractStatus.class);
+        Assertions.assertThat(timeToLiveIndexExists(indexOperations)).isTrue();
+    }
+
+    @Test
+    @Disabled("Long running test that depends on external TTL config, enable when needed")
+    public void When_TimeToLiveHasPassedInEhrExtractStatusRepository_Expect_DocumentRemoved() {
```
Comment (mattd-kainos) [inline service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/TimeToLiveTest.java:49]: This test fails when enabled.

**Review by mattd-kainos — CHANGES_REQUESTED** [review]
mongo index on the wrong field, disabled test fails when enabled

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/TimeToLiveTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import org.springframework.data.mongodb.core.index.IndexInfo;
+import org.springframework.data.mongodb.core.index.IndexOperations;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+@ExtendWith({ SpringExtension.class, MongoDBExtension.class, ActiveMQExtension.class})
+@SpringBootTest
+@DirtiesContext
+public class TimeToLiveTest {
+    private static final int MAX_AWAIT_IN_SECONDS = 90;
```
Comment (mattd-kainos) [inline service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/TimeToLiveTest.java:32]: src/main/resources/application.yml defines the default TTL for the app. The story does specify it, but it should be 7 days.

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/TimeToLiveTest.java (hunk 2 above)
Comment (mattd-kainos) [reply service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/TimeToLiveTest.java:49]: If we make the default in src/intTest/resources/application.yml much shorter (a few seconds):
1) could we have this test always run without our integration tests taking way too long to finish
2) do the other repo intTest still run reliably?

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/TimeToLiveTest.java (hunk 3 above)
Comment (ElliotB-96) [reply service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/TimeToLiveTest.java:32]: The test fails when running below 40 seconds, leaving at 90 at discussed.

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/TimeToLiveTest.java (hunk 2 above)
Comment (ElliotB-96) [reply service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/TimeToLiveTest.java:49]: Tests were running and passing with the config set to 2 and 3 seconds, left disabled as tests fail when MAX_AWAIT_IN_SECONDS is set to less than 40 seconds, left disabled as discussed.

File: service/src/intTest/resources/application.yml [hunk 4]
```diff
@@ -35,5 +35,5 @@ gp2gp:
     password: ${GP2GP_MONGO_PASSWORD:}
     options: ${GP2GP_MONGO_OPTIONS:}
     autoIndexCreation: ${GP2GP_MONGO_AUTO_INDEX_CREATION:true}
-    ttl: ${GP2GP_MONGO_TTL:P30D}
+    ttl: ${GP2GP_MONGO_TTL:PT3S}
```
Comment (mattd-kainos) [inline service/src/intTest/resources/application.yml:38]: We need to update src/main/resources/application.yml to have a default of 7 days

**Review by mattd-kainos — DISMISSED** [review]

**Review by Broom94 — APPROVED** [review]

**Review by mattd-kainos — APPROVED** [review]

**Pull request merged**